### PR TITLE
Support Managed Identity

### PR DESCRIPTION
**Pull request by YanaXu**
### Description
This PR is going to support both system- and user- assigned managed identity login for self-hosted runners on Azure VM. 

#### What's new
- The Action provides a parameter `auth-type` with value list `[SERVICE_PRINCIPAL, IDENTITY]` to identify the type of authentication.
  1. If `auth-type: SERVICE_PRINCIPAL` with `clientId`, `tenantId` and `clientSecret` detected in your input, we will attempt to login by using service principal with the secret.
  ```yaml
  #login with secret
  - uses: azure/login@v1
        with:
          auth-type: SERVICE_PRINCIPAL
          creds: ${{ secrets.AZURE_CREDENTIALS }}
  ```
  2. If `auth-type: SERVICE_PRINCIPAL` with `clientId` and `tenantId` detected in your input, we will attempt to login by using OIDC.
  ```yaml
  #login with OIDC
  - uses: azure/login@v1
      with:
          auth-type: SERVICE_PRINCIPAL
          client-id: ${{ secrets.AZURE_CLIENT_ID }}
          tenant-id: ${{ secrets.AZURE_TENANT_ID }}
          subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
  ```
  3. If `auth-type: IDENTITY` with `clientId` detected in your input, we will attempt to login by using user-assigned managed identity.
   ```yaml
   #login with user-assigned managed identity
  - uses: azure/login@v1
      with:
          auth-type: IDENTITY
          client-id: ${{ secrets.AZURE_CLIENT_ID }}
          subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
  ```
  4. If `auth-type: IDENTITY` without `clientId` detected in your input, we will attempt to login by using system-assigned managed identity.
   ```yaml
   #login with system-assigned managed identity
  - uses: azure/login@v1
    with:
          auth-type: IDENTITY
          subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
  ```
The default value of `auth-type` to be `SERVICE_PRINCIPAL` to make it backward-compatible.
- Always specify the `subscription-id` for safety. Without doing so will make Azure CLI use the current active subscription, which can be risky if your default subscription has changed.

- The changes for README.md is in https://github.com/Azure/login/pull/344.

### CODE REVIEW

**Comment by YanaXu** [comment]
Hi @jiasli , @isra-fel , @dolauli , please help us review this PR. 
This design spec is [Design Spec for Supporting Managed Identity in Azure Login Action.docx](https://microsoftapc-my.sharepoint.com/:w:/g/personal/yanxu1_microsoft_com/EaC8h4-0_ZdOh0xN1CJhhwwB335SXuZjRo7FdcUcDoaukw?e=p9hBPc). 
Please refer to [this workflow file](https://github.com/Azure/login/actions/runs/6335360141/workflow) for the configuration of this action to support difference login methods.

- The Azure CLI login commands are in [src/cli/AzureCliLogin.ts](https://github.com/Azure/login/blob/yanxu/mi/src/Cli/AzureCliLogin.ts).
- The Azure PowerShell login cmdlets are in [src/PowerShell/AzPSScriptBuilder.ts](https://github.com/Azure/login/blob/yanxu/mi/src/PowerShell/AzPSScriptBuilder.ts).

The readme.md of this action will be updated separately in a later PR.

File: __tests__/PowerShell/AzPSScriptBuilder.test.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+    test('PS script should not set context while passing allowNoSubscriptionsLogin as true', () => {
+        setEnv('environment', 'azurecloud');
+        setEnv('enable-AzPSSession', 'true');
+        setEnv('allow-no-subscriptions', 'true');
+        setEnv('auth-type', 'SERVICE_PRINCIPAL');
+        let creds = {
+            'clientId': 'client-id',
+            'clientSecret': 'client-secret',
+            'tenantId': 'tenanat-id',
```
Comment (MoChilia) [inline __tests__/PowerShell/AzPSScriptBuilder.test.ts:34]: A typo: `tenanat-id` -> `tenant-id` in this file.

File: src/PowerShell/AzPSLogin.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                    if (error && error.trim().length !== 0) {
+                        commandStdErr = true;
+                        core.error(error);
+                    }
+                }
+            }
+        };
+
+        let psPath:string = await io.which(AzPSConstants.PowerShell_CmdName, true);
+        await exec.exec(`"${psPath}" -Command`, [psScript], options)
```
Comment (MoChilia) [inline src/PowerShell/AzPSLogin.ts:83]:
```suggestion
        await exec.exec(`"${psPath}"`, [' -Command', psScript], options)
```
Let's move the argument `-Command` to `args`.

File: src/PowerShell/AzPSScriptBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            }
+        }
+
+        if (!loginConfig.allowNoSubscriptionsLogin && loginConfig.subscriptionId) {
+            if (loginConfig.tenantId) {
+                commands += `Set-AzContext -SubscriptionId '${loginConfig.subscriptionId}' -TenantId '${loginConfig.tenantId}' | out-null;`;
+            } else {
+                commands += `Set-AzContext -SubscriptionId '${loginConfig.subscriptionId}' | out-null;`;
+            }
+        }
```
Comment (MoChilia) [inline src/PowerShell/AzPSScriptBuilder.ts:58]: Shall we extract this to a function called `setSubscription` and add a warning message when the arguments `allow-no-subscriptions` and `subscription-id` are both missed?

Comment (YanaXu) [reply src/PowerShell/AzPSScriptBuilder.ts]: - Extract to a new function: done.
- Add a warning message: I think it should be added in **LoginConfig**.

File: __tests__/PowerShell/AzPSScriptBuilder.test.ts (hunk 1 above)
Comment (YanaXu) [reply __tests__/PowerShell/AzPSScriptBuilder.test.ts]: Thanks. It's updated.

File: src/PowerShell/AzPSLogin.ts (hunk 2 above)
Comment (YanaXu) [reply src/PowerShell/AzPSLogin.ts]: Done.

File: src/Cli/AzureCliLogin.ts [hunk 4]
```diff
@@ -86,22 +88,67 @@ export class AzureCliLogin {
...
+        core.info(`Azure CLI login succeeds by using ${methodName}.`);
+    }
 
-        console.log(`Done registering cloud: "${this.loginConfig.environment}"`)
+    async setSubscription() {
+        if (this.loginConfig.allowNoSubscriptionsLogin) {
+            return;
+        }
+        if (!this.loginConfig.subscriptionId) {
+            core.warning('No subscription-id is given. Skip setting subscription... If there are mutiple subscriptions under the tenant, please input subscription-id to specify which subscription to use.');
```
Comment (jiasli) [inline src/Cli/AzureCliLogin.ts:138]: > No subscription-id is given. Skip setting subscription...

This statement is not accurate. Azure CLI chooses the first active subscription as the default/subscription, so a implicit setting subscription is performed.

File: src/common/LoginConfig.ts [hunk 5]
```diff
@@ -9,78 +9,81 @@ export class LoginConfig {
...
     }
 
     async validate() {
-        if (!this.servicePrincipalId || !this.tenantId || !(this.servicePrincipalKey || this.enableOIDC)) {
-            throw new Error("Not all values are present in the credentials. Ensure clientId, clientSecret and tenantId are supplied.");
+        if (!LoginConfig.azureSupportedCloudName.has(this.environment)) {
+            throw new Error("Unsupported value for environment is passed. The list of supported values for environment are 'azureusgovernment', 'azurechinacloud', 'azuregermancloud', 'azurecloud' or 'azurestack'");
         }
-        if (!this.subscriptionId && !this.allowNoSubscriptionsLogin) {
-            throw new Error("Not all values are present in the credentials. Ensure subscriptionId is supplied.");
```
Comment (jiasli) [inline src/common/LoginConfig.ts:80]: The original design makes more sense to me. Without requiring the `subscriptionId`, the user may unexpectedly perform operations in a wrong subscription.

If this feature is not related to managed identity, I suggest we split this change to another PR.

File: src/Cli/AzureCliLogin.ts [hunk 6]
```diff
@@ -24,39 +30,35 @@ export class AzureCliLogin {
...
-                "--subscription",
-                this.loginConfig.subscriptionId
-            ];
-            await this.executeAzCliCommand(`account set`, true, loginOptions, args);
+            let args = ["--identity"];
+            if (this.loginConfig.servicePrincipalId) {
+                await this.loginWithUserAssignedIdentity(args);
+            }
+            else {
+                await this.loginWithSystemAssignedIdentity(args);
```
Comment (jiasli) [inline src/Cli/AzureCliLogin.ts:60]: Can `loginWithSystemAssignedIdentity` function be written inline as it only contains 1 line and is called only once?

File: src/PowerShell/AzPSLogin.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        if (commandStdErr) {
+            throw new Error('Azure PowerShell login failed with errors.');
+        }
+        const result: any = JSON.parse(outputString.trim());
+        if (!(AzPSConstants.Success in result)) {
+            throw new Error(`Azure PowerShell login failed with error: ${result[AzPSConstants.Error]}`);
+        }
+        return result[AzPSConstants.Result];
+    }
+}
```
Comment (jiasli) [inline src/PowerShell/AzPSLogin.ts:93]: This file lacks the ending new line.

File: src/Cli/AzureCliLogin.ts [hunk 8]
```diff
@@ -7,14 +7,20 @@ import * as io from '@actions/io';
 export class AzureCliLogin {
     loginConfig: LoginConfig;
     azPath: string;
-    
+    loginOptions: ExecOptions;
```
Comment (jiasli) [inline src/Cli/AzureCliLogin.ts:10]: I am not seeing `loginOptions` being used when instantiating `AzureCliLogin`. Why is it put in the `constructor`?

**Comment by jiasli** [comment]
> In the previous code, `creds` is not compatible with individual parameters. We use `creds` for login using service principal with secret and individual parameters for `OIDC` login:

I personally prefer the original design as it is explicit and definitive. The new design allows using OIDC with either individual parameters or `creds`.

**Comment by YanaXu** [comment]
Hi @MoChilia , @jiasli and @isra-fel , 
This PR is updated based on the review comments. Could you please review it again?

These comments are applied:
- 'creds' will only support login with SP+secrets.
- 'creds' will only be valid when there are no individual parameters and 'auth-type' is 'SERVICE_PRINCIPAL'.
- if 'allow-no-subscriptions' is not set or set to false while 'subscription-id' is not set, throw an error (it's same as before).
- In Azure PS login, don't use "Set-AzContext", add tenant id and subscription id in "Connect-AzAccount" instead.
- Import Az.Accounts instead of adding its path to PSModulePath and print used Az.Accounts version.

This comment is skipped in this PR:
- Add a warning message if 'allow-no-subscriptions' is true. 

I also add the unit tests for parameter validation and Azure PS script generation. You can take them for reference.

**Review by msJinLei — APPROVED** [review]

**Review by MoChilia — APPROVED** [review]
Common and Azure CLI part LGTM.

File: src/PowerShell/AzPSScriptBuilder.ts [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    static async getAzPSLoginScript(loginConfig: LoginConfig) {
+        let loginMethodName = "";
+        let commands = 'Clear-AzContext -Scope Process; ';
+        commands += 'Clear-AzContext -Scope CurrentUser -Force -ErrorAction SilentlyContinue; ';
+
+        if (loginConfig.environment.toLowerCase() == "azurestack") {
+            commands += `Add-AzEnvironment -Name '${loginConfig.environment}' -ARMEndpoint '${loginConfig.resourceManagerEndpointUrl}' | out-null;`;
+        }
+        if (loginConfig.authType === LoginConfig.AUTH_TYPE_SERVICE_PRINCIPAL) {
+            if (loginConfig.servicePrincipalKey) {
```
Comment (isra-fel) [inline src/PowerShell/AzPSScriptBuilder.ts:33]: It seems the nested if-else can be refactored to something like `switch (loginConfig.GetLoginFlow()) { ... }` but I don't have a strong opinion here.

File: src/common/LoginConfig.ts [hunk 10]
```diff
@@ -2,85 +2,111 @@ import * as core from '@actions/core';
...
         this.tenantId = core.getInput('tenant-id', { required: false });
         this.subscriptionId = core.getInput('subscription-id', { required: false });
 
+        this.readParametersFromCreds();
+
         this.audience = core.getInput('audience', { required: false });
         this.federatedToken = null;
+
+        this.mask(this.servicePrincipalId);
+        this.mask(this.servicePrincipalKey);
```
Comment (isra-fel) [inline src/common/LoginConfig.ts:47]: Does servicePrincipalKey actually mean servicePrincipal secret?

File: action.yml [hunk 11]
```diff
@@ -1,6 +1,6 @@
 # Login to Azure subscription
 name: 'Azure Login'
-description: 'Authenticate to Azure using OIDC and run your Az CLI or Az PowerShell based actions or scripts. github.com/Azure/Actions'
+description: 'Authenticate to Azure using OIDC and run your Azure CLI or Azure PowerShell based actions or scripts. github.com/Azure/Actions'
```
[the remark this answers is not in the document]
Comment (YanaXu) [reply action.yml]: done.

File: src/PowerShell/AzPSLogin.ts [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        core.debug(`Azure PowerShell Login Script: ${loginScript}`);
+        await AzPSLogin.runPSScript(loginScript);
+        console.log(`Running Azure PowerShell Login successfully.`);
+    }
+
+    setPSModulePathForGitHubRunner() {
+        const runner: string = process.env.RUNNER_OS || os.type();
+        switch (runner.toLowerCase()) {
+            case "linux":
+                this.pushPSModulePath(AzPSConstants.DEFAULT_AZ_PATH_ON_LINUX);
```
[the remark this answers is not in the document]
Comment (YanaXu) [reply src/PowerShell/AzPSLogin.ts:40]: Yes. It'll also work on a self-hosted runner. 
There are 2 options for this:

- Only apply this for GitHub-Hosted runners. 
  - I see there is an env property **RUNNER_ENVIRONMENT** which have 2 values "self-hosted" and "github-hosted". But I didn't find official docs for it. That's why I don't suggest to use it in our code since it's not reliable. I didn't find other ways to check if this action is running on a GitHub-Hosted runner.
- The other way is to keep it as it is now.
  - If so, this line will be applied for "self-hosted" runner too. It could add an invalid path to PSModulePath. However, it does no harm. We have test cases for it.

I suggest we choose option 2.

File: src/PowerShell/AzPSLogin.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+                }
+            }
+        };
+
+        let psPath: string = await io.which(AzPSConstants.PowerShell_CmdName, true);
+        await exec.exec(`"${psPath}"`, ["-Command", psScript], options)
+        if (commandStdErr) {
+            throw new Error('Azure PowerShell login failed with errors.');
+        }
+        const result: any = JSON.parse(outputString.trim());
```
[the remark this answers is not in the document]
Comment (YanaXu) [reply src/PowerShell/AzPSLogin.ts]: done.

File: src/PowerShell/AzPSScriptBuilder.ts [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+import { LoginConfig } from '../common/LoginConfig';
+
+export default class AzPSScriptBuilder {
+
+    static getImportLatestModuleScript(moduleName: string): string {
+        let script = `try {
+            $ErrorActionPreference = "Stop"
+            $WarningPreference = "SilentlyContinue"
+            $output = @{}
+            $latestModulePath = (Get-Module -Name '${moduleName}' -ListAvailable | Sort-Object Version -Descending | Select-Object -First 1).Path
```
[the remark this answers is not in the document]
Comment (YanaXu) [reply src/PowerShell/AzPSScriptBuilder.ts:11]: This `${moduleName}` will be calculated by JS and passed to PS later. For example, if `moduleName `value is "Az.Accounts", when PS execute the script, it's `'Az.Accounts'`.

File: src/common/LoginConfig.ts (hunk 10 above)
Comment (YanaXu) [reply src/common/LoginConfig.ts]: yes. Renamed `servicePrincipalKey` to `servicePrincipalSecret`.

**Comment by anton-delphai** [comment]
It seems like this change makes `allow-no-subscriptions` option useless as `readParametersFromCreds()` already throws an error without considering this option

**Comment by pawan-regoti** [comment]
a funny workaround is giving not null value to `subscriptionId` in creds and set `allow-no-subscriptions` to `true`. It will bypass the subscriptionId and not throws error saying `subscriptionId` is missing.

```
     - uses: azure/login@v1
        with:
          allow-no-subscriptions: true
          ...
          creds: '{"subscriptionId":"something"}'

```
